examples/allocation/zig/testdata/greet.zig

```diff
@@ -0,0 +1,79 @@
+const std = @import("std");
+const allocator = std.heap.page_allocator;
+
+// main is required for Zig to compile to enable WASI.
+pub fn main() anyerror!void {
+    return;
+}
+
+extern "env" fn log(ptr: u32, size: u32) void;
+
+// _log prints a message to the console using log.
+pub fn _log(message: []u8) void {
+    const ptrSize = stringToPtr(message);
+    var p = ptrSize >> 32;
+    log(@truncate(u32, p), @truncate(u32, ptrSize));
+}
+
+pub export fn malloc(size: usize) usize {
+    var memory = allocator.alloc(u8, size) catch unreachable;
+    return @ptrToInt(memory.ptr);
+}
+
+pub export fn free(ptr: usize) void {
+    var s: []u8 = undefined;
+    s.ptr = @intToPtr([*]u8, ptr);
+    allocator.free(s);
+}
```

Comment:
Not sure it works properly...

Reply:
@Luukdegram @inkeliz can either of you do a pass on this to make sure we are suggesting good behavior for externalizing zig allocation functions in wasm?

Reply:
I _think_ this needs to set the size of []u8 as well..

Reply:
As mathetake mentioned, the size on the slice is missing. 
Here's a godbolt with an idiomatic implementation of malloc/free: https://godbolt.org/z/M7vWM6rhc

Some notes: You shouldn't really have to use `@ptrToInt` at all, pointers are nothing more than an offset into memory under the hood. By using `@ptrToInt` and `@intToPtr` you're doing what the compiler already does for you, yet removing all type information for the compiler which then cannot apply optimizations to it.
This is why you should use `?[*]u8` which represents an optional many-item pointer. This can be assigned `null` (really just offset 0 in Wasm), in case of an error, as you'd expect from `malloc`.

Reply:
Thanks for reviewing, all! I assumed that the allocated size was stored next to the heap pointer (technically wasm memory in this case) and I didn't have to pass the length.
Also, it is super cool to take `[*]u8` as an argument. It is much simpler. I've fixed it according to your comment. Thanks.
https://github.com/tetratelabs/wazero/pull/741/commits/ea5fa0ec27dec0e2ec1a7f0f87f2123ecd1f429f

Reply:
One more question. If `alloc` in Zig returns an error, `?[*]u8` can be null. In that case, will the return value (`namePtr`) be 0 in Go? Should we return an error if `namePtr` is 0?
https://github.com/tetratelabs/wazero/blob/b191b90928b9de4c53cd4f485daf0d71658fcba2/examples/allocation/zig/greet.go#L86

I'm wondering if it could return 0 even though it exits successfully.

Reply:
Yes, the return value (`namePtr`) will be 0 in Go.

Reply:
Do you mean we should add this sort of error handling?

```
if namePtr == 0 {
    return fmt.Errorf("malloc failed for %s", name)
}
```

Reply:
Sure. As allocations can fail, we would set the right example by handling such case.

Reply:
Thanks for confirming. I was concerned that 0 could be returned as the valid offset. Added.
https://github.com/tetratelabs/wazero/pull/741/commits/e5917dc7271d1cd919e788f59cbe01aa110f2a2a